src/platform/plugins/shared/cps/public/services/access_control.ts

```diff
@@ -0,0 +1,123 @@
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { ProjectRoutingAccess } from '@kbn/cps-utils';
+import { i18n } from '@kbn/i18n';
+
+/**
+ * Rule for determining access based on route pattern
+ */
+export interface RouteAccessRule {
+  /** Regex pattern to match against location hash */
+  pattern: RegExp;
+  /** Access level to grant when pattern matches */
+  access: ProjectRoutingAccess;
+}
+
+/**
+ * Configuration for a single app's access control
+ */
+export interface AppAccessConfig {
+  /** Default access level for this app (when no route rules match) */
+  defaultAccess: ProjectRoutingAccess;
+  /** Optional route-specific rules (checked in order) */
+  routeRules?: RouteAccessRule[];
+  /** Optional custom readonly message for this app */
+  readonlyMessage?: string;
+}
+
+/**
+ * Complete access control configuration
+ * Maps app IDs to their access configurations
+ */
+export type AccessControlConfig = Record<string, AppAccessConfig>;
+
+/**
+ * Default access control configuration
+ *
+ * Access Levels:
+ * - EDITABLE: Full functionality - users can change project scope
+ * - READONLY: View-only mode - shows current scope but prevents changes
+ * - DISABLED: Picker is completely disabled
+ *
+ * Default Configuration:
+ *
+ * | App        | Route                  | Access Level | Notes                              |
+ * |------------|------------------------|--------------|-------------------------------------|
+ * | dashboards | all routes except list | EDITABLE     | All dashboard pages except listing  |
+ * | dashboards | #/list                 | DISABLED     | List page only                     |
+ * | discover   | all routes             | EDITABLE     | All discover pages                 |
+ * | visualize  | type:vega in route     | EDITABLE     | Vega visualizations only           |
+ * | visualize  | other routes           | DISABLED     | Other visualization types          |
+ * | lens       | all routes             | READONLY     | All lens pages (read-only)         |
+ * | maps       | all routes             | EDITABLE     | All maps pages (read-only)         |
+ * | all others | all routes             | DISABLED     | Default behavior for unknown apps  |
+ */
+export const DEFAULT_ACCESS_CONTROL_CONFIG: AccessControlConfig = {
+  dashboards: {
+    defaultAccess: ProjectRoutingAccess.EDITABLE,
+    routeRules: [
+      {
+        pattern: /#\/list/,
+        access: ProjectRoutingAccess.DISABLED,
+      },
+    ],
+  },
+  discover: {
+    defaultAccess: ProjectRoutingAccess.EDITABLE,
+  },
+  visualize: {
+    defaultAccess: ProjectRoutingAccess.DISABLED,
+    routeRules: [
+      {
+        pattern: /type:vega/, // Restrict to only Vega visualizations
```

Comment:
I couldn't see this option in Serverless, but I thought it was safe to double-check: 

This URL-pattern check wouldn't work if `state:storeInSessionStorage` is enabled, correct?